File: 876. Middle of the Linked List/876. Middle of the Linked List.cpp
// Simple fast and slow pointer logic can give us the value over here that's it

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* middleNode(ListNode* head) {
        ListNode *prev,*fast,*slow;
        if(head==NULL or head->next==NULL){
            return head;
        }
        fast=head;
        slow=head;
        while(fast!=NULL and fast->next!=NULL){
            prev=slow;
            slow=slow->next;
            fast=fast->next->next;
        }
        return slow;
    }
};

File: 876. Middle of the Linked List/876. Middle of the Linked List.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


def middleNode(head):
    # Two pointers - slow and fast pointers
    slow, fast = head, head

    while fast.next and fast.next.next:
        slow = slow.next
        fast = fast.next.next

    # containts even number of elements
    return slow.next if fast.next!=None else slow
        